### ISSUE TITLE
Indentation based syntax and other questions

### ISSUE THREAD

**Issue by whitetigle** [ikigai-go/compiler#2]
Honestly after all these years of programming, I don't want to live anymore in a world where roads finish with a semi colon and are surrounded by invasive curly brackets 😉 

Please, please, please! 
A thousand thanks to not make this happen  🙏 🙏🙏🙏🙏🙏🙏🙏🙏🙏

**Comment by alfonsogarciacaro** [comment]
You mean indentation-based syntax, like F#/Python? Hmm, it seems it's possible to [do it with chevrotain](https://sap.github.io/chevrotain/docs/guide/custom_token_patterns.html) but honestly I'd prefer to stick with JS/TS-like syntax. There are several advantages for creating a quick prototype.

- Saves a lot of discussions ;)
- Much simpler to implement because we can basically ignore whitespace.
- It will feel less foreign for JS developers which may increase the chances of getting contributors. I personally think syntax is not the most important part of a language but I know it's actually one of the most difficult things to jump from one language to another. I'm afraid if we start creating indentation-bases syntax people will think Ikigai is yet-another-exotic-language-compiling to JS.

It's true at the beginning I wanted to use newlines instead of semicolons (but not auto-semicolon insertion as JS does), but I discarded it because of the complexity of whitespace tokenization. Another possibility I thought of is to offer an alternative syntax when the language is more mature (just by replacing the parser) and see what's the feedback.

What do you think? cc @MangelMaxime

**Comment by MangelMaxime** [comment]
Yes chevrotain is able to parse indentation-based syntax [Python Lexer](https://github.com/SAP/chevrotain/blob/master/examples/lexer/python_indentation/python_indentation.js).

I personally prefer lighter syntax. It forces you to structure your code correctly. I think Reason switched from indentation based syntax to "semi-colon" based syntax probably because they wanted to attract more people in the language.

But I find the code less readable than F# in comparison. Elm is using indentation based syntax and people are fine with it so I don't think we should just choose "semi-colon" just to attract more people. Users will come for the features first I think and with good documentation, the learning curve should be fine.

**Comment by MangelMaxime** [comment]
> I thought of is to offer an alternative syntax when the language is more mature (just by replacing the parser) and see what's the feedback.

We all know that the feedback will be people choosing what they are used too. Because of fear of the changes. Also, people will probably use the default syntax. Because the compiler will come with a default one activated and because it's easier to use compare to activate the other syntax with a flag then it will be more used.

I mentioned Reason example before, but I think F# did do the opposite move. It's was kind of "semi-colon" based at first then they went pure indentation based.

**Comment by whitetigle** [comment]
Anyway, I did not mean to disturb the process. 
So do whatever you want to but still if in the end we could avoid that I would be so happy 🦄 🌈 😄

**Comment by alfonsogarciacaro** [comment]
If I'm not mistaken, F# had at the beginning a syntax very similar to OCaml. OCaml doesn't use indentation but keywords to define the scopes `begin .. end`. With bindings you must always use `in`: `let foo = x in ..`.

F# added later a lighter syntax based on indentation that became the default. ReasonML has just replaced `in` with `;` but I suspect both tokens have the same function in their lexer.

Anyways, if we take this road this is the moment to decide it :) It's true that using JS syntax with statements would cause frictions in a primarily expression-based language (as Ikigai aspires to be) so a mini-F# like @MangelMaxime had in mind also has advantages :)

In that case, some likely differences with F# would be:

- No variable shadowing in same scope
- No implicitly curried functions: this creates lots of issues when interacting with JS
- No custom operators: they're not trivial and Elm is also discarding them now
- Bindings don't create scope
- `let | let mutable` for bindings or `const | mutable`?
- Lambdas à la F# `fun x -> ...` or C#/JS `x => ...`?
- Commas to separate collection items `[1,2,3]`. Tuples as arrays (like in TS) or with (mandatory) parens?
- Should immutable list have native syntax?

**Comment by MangelMaxime** [comment]
I will vote for the mini-F# like language :trollface: 



> * No variable shadowing in same scope

Ok, for now. We will see if this create friction or no.

> * No implicitly curried functions: this creates lots of issues when interacting with JS

Agree as we want to make interop with JS a breath 

> * No custom operators: they're not trivial and Elm is also discarding them now

I agree with that agree too :)

> * Bindings don't create scope
> * `let | let mutable` for bindings or `const | mutable`?

I think not all JavaScript developers understand what `const` are. And because we plan to make it stickier (not allowing mutating a field in a `const` unless marked as `mutable`) I am in favour of using F# `let | let mutable`.

> * Lambdas à la F# `fun x -> ...` or C#/JS `x => ...`?

When looking at the current situation of F#, there is traction to support `x => ...` syntax. So I think to make sense to the C#/JS syntax.

For the record, I am not in favour of adding this syntax in F# because it would make a mix of syntax in it for just removing 3 letters. But if we have it natively I am totally ok with it :)

> * Commas to separate collection items `[1,2,3]`. Tuples as arrays (like in TS) or with (mandatory) 
parens?

I always found using `;` separator for list and `,` for tuple confusing. I am in favour of using `,` everywhere and when defining a tuple we make the `()` mandatory. *I think Elm does that*

The only drawback about making `()` mandatory is if we have an Elmish like framework. We will need to write `(model, Cmd.none)`. It's a bit more verbose, but at least it's consistent everywhere.

And it's still easy to read:

```elm
        CompletedLogin (Err error) ->
            let
                serverErrors =
                    Api.decodeErrors error
                        |> List.map ServerError
            in
            ( { model | problems = List.append model.problems serverErrors }
            , Cmd.none
            )

        CompletedLogin (Ok viewer) ->
            ( model
            , Viewer.store viewer
            )
```
[Source](https://github.com/rtfeldman/elm-spa-example/blob/master/src/Page/Login.elm#L159-L198)

> * Should immutable list have native syntax?

I think we wanted to go immutable first. So for me `[1, 2]` is an immutable list, am I wrong?

**Comment by whitetigle** [comment]
Oh well @MangelMaxime just read into my mind to answer so I 100% agree with him. Guys you should stop doing this 😉 

> No variable shadowing in same scope 

ok

> No implicitly curried functions: this creates lots of issues when interacting with JS

ok

> No custom operators: they're not trivial and Elm is also discarding them now

ok

> Bindings don't create scope

ok

> let | let mutable for bindings or const | mutable?

let

> Lambdas à la F# fun x -> ... or C#/JS x => ...?

whatever is fine. But Maxime's arguments make sense.

> Commas to separate collection items [1,2,3]. Tuples as arrays (like in TS) or with (mandatory) parens?

Yes

> Should immutable list have native syntax?

immutable first by I will check your answer to that.
